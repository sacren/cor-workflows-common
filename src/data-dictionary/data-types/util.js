import _ from 'lodash'
import fp from 'lodash/fp'
import * as returnTypes from '../return-types'

// Takes an existing mapping of coercions (e.g. user.js COERCIONS constant) and returns
// a mapping where the keys are corresponding to the list of that singular type (e.g. text -> text-list),
// where the coercion functions for the singular type are wrapped in a map, such that any list of singular
// typed values will be each be converted using the underlying singular type coercion function.
//
// For example:
//
// If your coercion map looked like this:
//
// export const COERCIONS = {
//   [TEXT]: _.identity,
//   [NUMBER]: value => +value
// }
//
// And you called wrapCoercionsForListType(COERCIONS), you would get the following:
//
// export const COERCIONS_FOR_LIST_TYPE = {
//   [TEXT_LIST]: fp.map(COERCIONS[TEXT]),
//   [NUMBER_LIST]: fp.map(COERCIONS[NUMBER])
// }
//
// The key here is we never want to do something "special" for converting singular values that are wrapped
// in the list type, it should always delegate to the singular type coercion function for each value
// in the list.
//
// Note: If you attempt to wrap a singular base type which converts to a data type that does _not_ have
// a list counterpart data type, this function will throw an error.
//
export function wrapCoercionsForListType (singularTypeCoercions) {
  return _.reduce(
    singularTypeCoercions,
    (accumulator, coercionFn, rightSideType) => {
      const listDataTypeKey = _.toUpper(`${rightSideType}_LIST`)
      const listDataType = _.get(returnTypes, listDataTypeKey) || throw new Error(`List data type ${listDataTypeKey} is not supported for singular base type ${rightSideType} (see ./return-types.js)`)
      return _.set(accumulator, listDataType, fp.map(coercionFn))
    },
    {}
  )
}
