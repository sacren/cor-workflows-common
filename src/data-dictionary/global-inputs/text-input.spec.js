/* Copyright Â© 2016 Kuali, Inc. - All Rights Reserved
 * You may use and modify this code under the terms of the Kuali, Inc.
 * Pre-Release License Agreement. You may not distribute it.
 *
 * You should have received a copy of the Kuali, Inc. Pre-Release License
 * Agreement with this file. If not, please write to license@kuali.co.
 */

import TextInput from './text-input'
import TestDeflation from '../../test/utils/deflation'
import { extractUnsupportedPreferredOperators } from '../test/util'

describe('TextInput', () => {
  let input
  beforeEach(() => {
    input = new TextInput()
  })

  it('should return the data from the deflated payload to inflate', async () => {
    const inflated = await TextInput.inflate(null, { data: 'deflated' })
    expect(inflated).toEqual('deflated')
  })

  it('should report itself as a leaf node', () => {
    expect(input.isLeaf()).toBeTruthy()
  })

  it('should return an empty list for children', async () => {
    const children = await input.getChildren()
    expect(children).toMatchObject([])
  })

  it('ensures preferredOperators are all valid', () => {
    expect(extractUnsupportedPreferredOperators(TextInput)).toEqual([])
  })

  describe('setData', () => {
    it('should set the name of the input to the data passed in', () => {
      input.setData('abcdefg')
      expect(input.data).toEqual('abcdefg')
      expect(input.name).toEqual('abcdefg')
    })

    it('should truncate the name to the first 12 characters of data', () => {
      input.setData('abcdefghijklmnop')
      expect(input.data).toEqual('abcdefghijklmnop')
      expect(input.name).toEqual('abcdefghijkl...')
    })

    it('should set a default name if no data was set', () => {
      input.setData()
      expect(input.data).toBeUndefined()
      expect(input.name).toEqual(TextInput.displayName)
    })
  })

  describe('getValue', () => {
    it('should call getValue on the parent', async () => {
      const parent = {
        getValue: jest.fn()
      }
      input.parent = parent
      const value = await input.getValue()
      expect(value)
      expect(parent.getValue).toHaveBeenCalledWith({})
    })

    it('should return data and update the value map', async () => {
      const map = {}
      input.setData('abcdefg')
      const value = await input.getValue(map)
      expect(value).toEqual('abcdefg')
      expect(map[TextInput.type]).toEqual('abcdefg')
    })
  })

  describe('Validation', () => {
    it('should throw an error if there is no data', () => {
      expect(() => {
        input.validate()
      }).toThrowErrorMatchingSnapshot()
    })

    it('should throw and error if data is not a string', () => {
      expect(() => {
        input.validate(1234)
      }).toThrowErrorMatchingSnapshot()
    })

    it('should not throw an error is a string is passed for validation', () => {
      expect(() => {
        input.validate('abcd')
      }).not.toThrow()
    })
  })

  TestDeflation(parent => new TextInput(parent, null, 'abcd'), {
    type: TextInput.type,
    name: 'abcd',
    requiresParent: false,
    data: 'abcd'
  })
})
