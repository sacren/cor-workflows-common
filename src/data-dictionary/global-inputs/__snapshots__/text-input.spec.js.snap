// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Validation should throw an error if there is no data 1`] = `"Cannot create a Text Input Context without a string"`;

exports[`TextInput Validation should throw and error if data is not a string 1`] = `"Cannot create a Text Input Context without a string"`;
